test/common/utils_assert.h

```diff
@@ -49,4 +50,73 @@ void AssertSameType( const T& /*x*/, const T& /*y*/ ) {}
 
+#if TEST_CUSTOM_ASSERTION_HANDLER_ENABLED
+
+#include "tbb/global_control.h"
+
+//! Check that expression x raises assertion failure with message containing given substring.
+/** Calls utils::SetCustomAssertionHandler to set utils::AssertionFailureHandler as a handler. */
+#define TEST_CUSTOM_ASSERTION_HANDLER(x, substr)                           \
+    {                                                                      \
+        auto default_handler = utils::SetCustomAssertionHandler();         \
+        const char* message = nullptr;                                     \
+        bool okay = false;                                                 \
+        try {                                                              \
+            x;                                                             \
+        } catch(utils::AssertionFailure a) {                               \
+            okay = true;                                                   \
+            message = a.message;                                           \
+        }                                                                  \
+        utils::CheckAssertionFailure(__LINE__, #x, okay, message, substr); \
+        utils::ResetAssertionHandler(default_handler);                     \
+    }
+
+namespace utils {
+
+//! Exception object that holds a message.
+struct AssertionFailure {
+    const char* message;
+    AssertionFailure( const char* filename, int line, const char* expression, const char* comment );
+};
+
+AssertionFailure::AssertionFailure( const char* filename, int line, const char* expression, const char* comment ) :
+    message(comment)
+{
+    REQUIRE_MESSAGE(filename, "missing filename");
+    REQUIRE_MESSAGE(0 < line, "line number must be positive");
+    // All of our current files have fewer than 4000 lines.
+    REQUIRE_MESSAGE(line < 5000, "dubiously high line number");
+    REQUIRE_MESSAGE(expression, "missing expression");
+}
+
+void AssertionFailureHandler(const char* filename, int line, const char* expression, const char* comment) {
+    throw AssertionFailure(filename, line, expression, comment);
+}
+
+tbb::assertion_handler_type SetCustomAssertionHandler() {
+    auto default_handler = tbb::set_assertion_handler(AssertionFailureHandler);
+    auto custom_handler = tbb::get_assertion_handler();
+    REQUIRE_MESSAGE(custom_handler == AssertionFailureHandler, "Custom assertion handler was not set.");
+    return default_handler;
+}
+
+void CheckAssertionFailure(int line, std::string expression, bool okay, const char* message, std::string substr) {
+    REQUIRE_MESSAGE(okay, "Line ", line, ", ", expression, " failed to fail");
+    REQUIRE_MESSAGE(message, "Line ", line, ", ", expression, " failed without a message");
+
+    std::string msg_str = message;
+    REQUIRE_MESSAGE(msg_str.find(substr) != std::string::npos,
+        expression, " failed with message '", msg_str, "' missing substring '", substr, "'");
```

Comment:
Why searching for a substring instead of requiring the exact match?

Reply:
I copied this over from TBB 2020. My guess is that it can help when, e.g., the assertion message includes a variable parameter.

Reply:
I think it does not matter, because the test uses public interfaces for assertion handling, and therefore sets the message content that should be thrown. So, it is up to the test what message to wait for in the fired assertion.

Reply:
Oh, okay. I see now. Thanks for the explanation. The assertion handler tests are sort of whitebox tests that require certain messages in the oneTBB asserts.